experimental/devices/ads1x15/ads1x15.go

```diff
@@ -0,0 +1,392 @@
+}
+
+func (d *Dev) prepareQuery(mux int, maxVoltage physic.ElectricPotential, minimumFrequency physic.Frequency) (pin AnalogPin, err error) {
+	// Determine the most appropriate gain
+	gain, err := d.bestGainForElectricPotential(maxVoltage)
+	if err != nil {
+		return
+	}
+
+	// Validate the gain.
+	gainConf, ok := d.gainConfig[gain]
+	if !ok {
+		err = errors.New("Gain must be one of: 2/3, 1, 2, 4, 8, 16")
+		return
+	}
+
+	// Determine the voltage multiplier for this gain
+	voltageMultiplier, ok := d.gainVoltage[gain]
+	if !ok {
+		err = errors.New("Gain must be one of: 2/3, 1, 2, 4, 8, 16")
+		return
+	}
+
+	// Determine the most appropriate data rate
+	dataRate, err := d.bestDataRateForFrequency(minimumFrequency)
+	if err != nil {
+		return
+	}
+
+	dataRateConf, ok := d.dataRates[dataRate]
+
+	if !ok {
+		// Write a nice error message in case the data rate is not found
+		keys := []int{}
+		for k := range d.dataRates {
+			keys = append(keys, k)
+		}
+
+		err = fmt.Errorf("Invalid data rate. Accepted values: %d", keys)
+		return
+	}
+
+	// Build the configuration value
+	var config uint16
+	config = ads1x15ConfigOsSingle // Go out of power-down mode for conversion.
+	// Specify mux value.
+	config |= uint16((mux & 0x07) << ads1x15ConfigMuxOffset)
+	// Validate the passed in gain and then set it in the config.
+	config |= gainConf
+	// Set the mode (continuous or single shot).
+	config |= ads1x15ConfigModeSingle
+
+	// Set the data rate (this is controlled by the subclass as it differs
+	// between ADS1015 and ADS1115).
+	config |= dataRateConf
+	config |= ads1x15ConfigCompQueDisable // Disable comparator mode.
+
+	// Build the query to the ADC
+	configBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(configBytes, config)
+	preparedQuery := append([]byte{ads1x15PointerConfig}, configBytes...)
+
+	// The wait for the ADC sample to finish is based on the sample rate plus a
+	// small offset to be sure (0.1 millisecond).
+	waitTime := time.Duration(1000000/dataRate+100) * time.Microsecond
+
+	pin = &ads1x15AnalogPin{
+		adc:               d,
+		preparedQuery:     preparedQuery,
+		voltageMultiplier: voltageMultiplier,
+		waitTime:          waitTime,
+	}
+
+	return
+}
+
+func (d *Dev) executePreparedQuery(query []byte, waitTime time.Duration, voltageMultiplier physic.ElectricPotential) (value physic.ElectricPotential, err error) {
+	// Lock the ADC converter to avoid multiple simultaneous readings.
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// Send the config value to start the ADC conversion.
+	// Explicitly break the 16-bit value down to a big endian pair of bytes.
+	if err = d.c.Tx(query, nil); err != nil {
+		return
+	}
+
+	// Wait for the ADC sample to finish.
+	time.Sleep(waitTime)
+
+	// Retrieve the result.
+	data := []byte{0, 0}
+	if err = d.c.Tx([]byte{ads1x15PointerConversion}, data); err != nil {
+		return
+	}
+
+	// Convert the raw data into physical value.
+	result := physic.ElectricPotential(binary.BigEndian.Uint16(data))
```

Comment:
I'd prefer something like:
```
raw := int16(binary.BigEndian.Uint16(data))
return physic.ElectricPotential(raw) * mult / physic.ElectricPotential(1<<15)
```

Especially that I see value to eventually just return the raw reading.
Or better (?) maybe the package analog should define an analog reading as:

```
// Reading is the result of DAC.Read()  (obviously not the case right now but this could be)
type Reading struct {
  V   physic.ElectricPotential
  Raw int32
}
```

then users are free to use the relevant value.

Reply:
Excellent idea, I've implemented it.